### PR TITLE
#792 - Initial commit of the sync_with_reality rework

### PR DESCRIPTION
**Pull request by kvanhijf**
All block-devices are now parsed making use of 'lsblk'

### CODE REVIEW

File: ovs/extensions/generic/disk.py [hunk 1]
```diff
@@ -27,128 +28,124 @@ class DiskTools(object):
     """
     This class contains various helper methods wrt Disk maintenance
     """
-    _logger = LogHandler.get('extensions', name='disktools')
+    _logger = LogHandler.get('extensions', name='disk-tools')
 
     @staticmethod
-    def create_partition(disk_path, disk_size, partition_start, partition_size):
+    def create_partition(disk_aliases, disk_size, partition_start, partition_size):
```
Comment (khenderick) [inline ovs/extensions/generic/disk.py]: You should pass in the alias you want to use

File: ovs/lib/disk.py [hunk 2]
```diff
@@ -43,198 +37,183 @@ class DiskController(object):
...
-        for storagerouter in storagerouters:
-            try:
-                client = SSHClient(storagerouter, username='root')
-            except UnableToConnectException:
-                DiskController._logger.info('Could not connect to StorageRouter {0}, skipping'.format(storagerouter.ip))
+        storagerouter = StorageRouter(storagerouter_guid)
+        try:
+            client = SSHClient(storagerouter, username='root')
+        except UnableToConnectException:
+            DiskController._logger.info('Could not connect to StorageRouter {0}'.format(storagerouter.ip))
```
Comment (khenderick) [inline ovs/lib/disk.py]: Don't catch the exception but let it fall through.

File: ovs/lib/disk.py [hunk 3]
```diff
@@ -43,198 +37,183 @@ class DiskController(object):
...
+        # Retrieve all symlinks for all devices
+        # Example of path_mapping:
+        # {'/dev/md0': ['/dev/disk/by-id/md-uuid-ad2de634:26d97253:5eda0a23:96986b76', '/dev/disk/by-id/md-name-OVS-1:0'],
+        #  '/dev/sda': ['/dev/disk/by-path/pci-0000:03:00.0-sas-0x5000c295fe2ff771-lun-0'],
+        #  '/dev/sda1': ['/dev/disk/by-uuid/e3e0bc62-4edc-4c6b-a6ce-1f39e8f27e41', '/dev/disk/by-path/pci-0000:03:00.0-sas-0x5000c295fe2ff771-lun-0-part1']}
+        name_path_mapping = {}
+        path_name_mapping = {}
+        for path_type in client.dir_list('/dev/disk'):
+            directory = '/dev/disk/{0}'.format(path_type)
+            if not client.dir_exists(directory=directory):
```
Comment (khenderick) [inline ovs/lib/disk.py]: It will exist, since dir_list reported it.

File: ovs/lib/disk.py [hunk 4]
```diff
@@ -43,198 +37,183 @@ class DiskController(object):
...
+        device_regex = re.compile('^KNAME="(?P<name>.*)" SIZE="(?P<size>\d*)" MODEL="(?P<model>.*)" STATE="(?P<state>.*)" MAJ:MIN="(?P<dev_nr>.*)" FSTYPE="(?P<fstype>.*)" TYPE="(?P<type>.*)" ROTA="(?P<rota>[0,1])" MOUNTPOINT="(?P<mtpt>.*)" PHY-SEC="(?P<sector_size>\d*)"$')
+        configuration = {}
+        parsed_devices = []
+        for device in devices:
+            match = re.match(device_regex, device)
+            if match is None:
+                DiskController._logger.error('Device regex did not match for {0}. Please investigate'.format(device))
+                raise Exception('Failed to parse \'lsblk\' output')
+
+            name = match.groupdict()['name'].strip()
```
Comment (khenderick) [inline ovs/lib/disk.py]: Execute `.groupdict()` once

File: ovs/lib/disk.py [hunk 5]
```diff
@@ -43,198 +37,183 @@ class DiskController(object):
...
-                    path = None
-                    for path_type in ['by-id', 'by-uuid']:
-                        if path is not None:
+            friendly_path = '/dev/{0}'.format(name)
+            system_aliases = name_path_mapping.get(friendly_path, [friendly_path])
+            DiskController._logger.info('Investigating device {0}'.format(friendly_path))
+
+            device_state = None
+            if client.file_read_link('/sys/block/{0}'.format(name)) is not None:  # If this returns, it means its a device and not a partition
+                device_state = 'OK' if state == 'running' or dev_nr.split(':')[0] != '8' else Disk.STATES.FAILURE
```
Comment (khenderick) [inline ovs/lib/disk.py]: Any other device types that need to be checked?

File: ovs/lib/disk.py [hunk 6]
```diff
@@ -43,198 +37,183 @@ class DiskController(object):
...
+            DiskController._logger.info('Could not connect to StorageRouter {0}'.format(storagerouter.ip))
+            return
+
+        # Retrieve all symlinks for all devices
+        # Example of path_mapping:
+        # {'/dev/md0': ['/dev/disk/by-id/md-uuid-ad2de634:26d97253:5eda0a23:96986b76', '/dev/disk/by-id/md-name-OVS-1:0'],
+        #  '/dev/sda': ['/dev/disk/by-path/pci-0000:03:00.0-sas-0x5000c295fe2ff771-lun-0'],
+        #  '/dev/sda1': ['/dev/disk/by-uuid/e3e0bc62-4edc-4c6b-a6ce-1f39e8f27e41', '/dev/disk/by-path/pci-0000:03:00.0-sas-0x5000c295fe2ff771-lun-0-part1']}
+        name_path_mapping = {}
+        path_name_mapping = {}
```
Comment (khenderick) [inline ovs/lib/disk.py]: Maybe call this `alias_name_...` and vice versa

File: ovs/lib/disk.py (hunk 5 above)
Comment (kvanhijf) [reply ovs/lib/disk.py]: NVME devices do NOT report state
LOOP devices do NOT report state
RAID devices do NOT report state
LVM devices do NOT report state

So for now sticking to major device number 8 only

**Review by khenderick — APPROVED** [review]

**Comment by kvanhijf** [comment]
https://github.com/openvstorage/framework/pull/1005  --> openvstorage-2.7.4-rev.4123.ef930a7
https://github.com/openvstorage/alba-asdmanager/pull/127 --> openvstorage-backend-1.7.4-rev.756.088c2ee
https://github.com/openvstorage/framework-alba-plugin/pull/231 --> openvstorage-sdm-1.6.4-rev.386.8590475

**Pull request merged**